### PR TITLE
fix to how we abort and retry connections

### PR DESCRIPTION
**Pull request by eundersander**
Taking another stab at this.

### CODE REVIEW

File: Assets/Scripts/NetworkClient.cs [hunk 1]
```diff
@@ -78,30 +93,39 @@ private IEnumerator TryConnectToServers()
...
+                    yield return new WaitForSeconds(4);
+                }
 
                 if (mainWebSocket == null)
                 {
-                    Debug.LogWarning($"Timeout connecting to {url}.");
-                    // Note that this websocket object is in the middle of an async
-                    // call to Connect(), but this class appears safe to call
-                    // CancelConnection in this situation.
-                    websocket.CancelConnection();
```
Comment (eundersander) [inline Assets/Scripts/NetworkClient.cs:104]: CancelConnection actually doesn't do anything useful for a socket still trying to connect. There is no way to cancel it. So I set a flag below and close it if/when it eventually gets opened.

File: Assets/Scripts/NetworkClient.cs [hunk 2]
```diff
@@ -113,10 +137,17 @@ private IEnumerator TryConnectToServers()
         }
     }
 
-    private async Task ConnectWebSocket(WebSocket websocket, string url)
+    private async Task ConnectWebSocket(WebSocket websocket, string url, FlagObject doAbort)
     {
         websocket.OnOpen += () =>
         {
+            if (doAbort.IsSet())
```
Comment (eundersander) [inline Assets/Scripts/NetworkClient.cs:144]: And just sanity-checking my understanding: there is absolutely no multi-threading or race-condition concern here. Async functions in Unity script get run later by Unity on the main Unity thread.

**Review by eundersander — COMMENTED** [review]
Some comments.

File: Assets/Scripts/NetworkClient.cs (hunk 2 above)
Comment (0mdc) [reply Assets/Scripts/NetworkClient.cs:144]: Yes, that's right.

**Review by 0mdc — APPROVED** [review]
LGTM!

**Pull request merged**